internal/resolve/service.go

```diff
@@ -4,61 +4,179 @@ import (
-	Pow       string          `json:"pow"`        // Proof of work
+}
+
+// OrganisationInfo is a structure returned by the external resolver system
+type OrganisationInfo struct {
+	Hash        string                        `json:"hash"`        // Hash of the organisation
+	PublicKey   bmcrypto.PubKey               `json:"public_key"`  // PublicKey of the organisation
+	Pow         string                        `json:"pow"`         // Proof of work
+	Validations []organisation.ValidationType `json:"validations"` // Validations for this organisation
 }
 
 // KeyRetrievalService initialises a key retrieval service.
 func KeyRetrievalService(repo Repository) *Service {
+	c, err := lru.New(64)
+	if err != nil {
+		c = nil
+	}
+
 	return &Service{
-		repo: repo,
+		repo:         repo,
+		routingCache: c,
+	}
+}
+
+// ResolveAddress resolves an address.
+func (s *Service) ResolveAddress(addr address.HashAddress) (*AddressInfo, error) {
+	logrus.Debugf("Resolving address %s", addr.String())
+	info, err := s.repo.ResolveAddress(addr)
+	if err != nil {
+		logrus.Debugf("Error while resolving address %s: %s", addr.String(), err)
+		return nil, err
+	}
+
+	// Resolve routing info, we often need it
+	ri, err := s.repo.ResolveRouting(info.RoutingID)
+	if err != nil {
+		logrus.Debugf("Error while resolving routing %s for address %s: %s", info.RoutingID, addr.String(), err)
+		return nil, err
 	}
+	info.RoutingInfo = *ri
+
+	return info, err
 }
 
-// Resolve resolves an address.
-func (s *Service) Resolve(addr address.HashAddress) (*Info, error) {
-	logrus.Debugf("Resolving %s", addr.String())
-	info, err := s.repo.Resolve(addr)
+// ResolveRouting resolves a route.
+func (s *Service) ResolveRouting(routingID string) (*RoutingInfo, error) {
+	// Fetch from cache if available
+	if s.routingCache != nil {
+		info, ok := s.routingCache.Get(routingID)
+		if ok {
+			logrus.Debugf("Resolving cached %s", routingID)
+			return info.(*RoutingInfo), nil
+		}
+	}
+
+	logrus.Debugf("Resolving %s", routingID)
+	info, err := s.repo.ResolveRouting(routingID)
 	if err != nil {
-		logrus.Debugf("Error while resolving %s: %s", addr.String(), err)
+		logrus.Debugf("Error while resolving route %s: %s", routingID, err)
+		return nil, err
+	}
+
+	// Store in cache if available
+	if s.routingCache != nil {
+		_ = s.routingCache.Add(routingID, info)
+	}
+	return info, nil
+}
+
+// ResolveOrganisation resolves a route.
+func (s *Service) ResolveOrganisation(orgHash address.HashOrganisation) (*OrganisationInfo, error) {
+	logrus.Debugf("Resolving %s", orgHash)
+	info, err := s.repo.ResolveOrganisation(orgHash)
+	if err != nil {
+		logrus.Debugf("Error while resolving organisation %s: %s", orgHash, err)
 	}
 
 	return info, err
 }
 
-// UploadInfo uploads resolve information to one (or more) resolvers
-func (s *Service) UploadInfo(info internal.AccountInfo) error {
+// UploadAddressInfo uploads resolve information to one (or more) resolvers
```

Comment:
We need the serialnumber to compute the signature?

Reply:
yes.. uploads should be changed.. it doesn't account for the serial number yet... I will add a commit for this

Reply:
fixed in the actual remote-repository instead of the service. This prevents leaking implementation details to the outside world.